### ISSUE TITLE
Consistency in Materialized Views

### ISSUE THREAD

**Issue by duarten** [scylladb/scylla#3340]
*Installation details*
Scylla version (or git commit hash): master

A base replica B*i* updates, in the common case of the pairing strategy, two view replicas V*ti* and V*ri*, which receive the view tombstone and the new view row, respectively. This strategy requires that nodes calculate view updates on every change to a base row, such that the previous view row can be correctly identified and invalidated. 

The following failure scenarios can cause the base and view tables to diverge:

- B*i* becomes unavailable after updating the view table: the update becomes a ghost record in the view, absent from the base table;
- V*ti* becomes unavailable after receiving the tombstone: the same base row will map to two view rows, one of them being a ghost record;
- V*ri* becomes unavailable after receiving the update: the base row will map to no rows in the view, or it will be outdated.

More than one node can become unavailable, and the consequences also vary with the request’s CL, range movements, and with subsequent repair operations, but there are generally two categories of inconsistency:

1. A base table update failed to propagate to the view;
2. The view contains a ghost record that maps to no base row with that value.

We can solve the first consistency issue by remembering a base table update for which we haven’t received successful acknowledgements from the view replicas. We can leverage hinted handoff for this. Eventually, a node will have to discard these remembered updates if it is unable to successfully propagate them.

To help solve the second consistency issue, and corner cases of the first issue, we can probabilistically repair a view against the base table. When we're reading from a view and trigger this repair, after we get the view rows and are able to calculate the matching base key, we’ll initiate a background read of the relevant rows from the paired base replica. This will let us know whether the view row is a ghost record, or whether the view row is not up to date.

In addition, we should have a full-fledged view repair process, which scans both base and view tables, and heals both types of inconsistencies. It's expensive to do a full scan on every such repair, so we would need some bookkeeping about repaired ranges.

**Comment by slivne** [comment]
created #3383 to use hintted handoff for base --> view failed updates

**Comment by nyh** [comment]
Right. I thought the last paragraph referred ("we should also have") to another issue, not that it adds another idea to this issue :-)

**Comment by duarten** [comment]
Inconsistencies between the base and the view are supposed to be rare. To occur, the user needs to be employing low consistency levels, and/or overloading the cluster such that view updates are dropped. There are some counters (such as dropped view updates) and events (such as nodes dying and data being lost) that suggest a repair is needed.

We can also provide a tool to detect inconsistencies between the base table and the view, based on table scans. This can be a basic script, expensive to run but accurate in the diagnosis. The user can then remedy the inconsistency by dropping and rebuilding the view. This tool can also be integrated into the Scylla Manager.

Note that even updating views in a transaction together with the base is not a guarantee that these tables won't become inconsistent. It is possible for the set of replicas (either base or view replicas) of a particular row to die at the same time, resulting in data loss. This is however a very low probability event, which makes the solution above more attractive due to it simplicity.
